### DISCUSSION TITLE
Kafka Extension implementation

### DISCUSSION THREAD

**Discussion by 3xc3ption** [ddev/ddev#3245] [General]
Hi,

i want to use an the kafka client (https://github.com/edenhill/librdkafka) in a ddev container as an extension. 
Unfortunately, there is no ddev-contrib.

Is it possible to implement the extension?

Beste regards
3xc3ption

**Comment by rfay** [comment]
I don't know much about how you'd use this, but to get it installed in the web container you just need a .ddev/web-build/Dockerfile with the installation instructions on https://github.com/edenhill/librdkafka

The below seems to work. Change amd64 to arm64 if you're on an arm64 machine.

```dockerfile
ARG BASE_IMAGE
FROM $BASE_IMAGE
RUN wget -qO - https://packages.confluent.io/deb/6.2/archive.key | sudo apt-key add -
RUN echo "deb [arch=amd64] https://packages.confluent.io/deb/6.2 stable main" >/etc/apt/sources.list.d/confluent.list
RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y -o Dpkg::Options::="--force-confold" --no-install-recommends --no-install-suggests librdkafka-dev
```

**Reply by 3xc3ption** [reply]
Hi rfay,

thank you very much. DDEV is so great. :-) My solution looks like so, i only need the client for kafka:

```php
ARG BASE_IMAGE
FROM $BASE_IMAGE
RUN sudo apt-get update -y && DEBIAN_FRONTEND=noninteractive apt-get install -y -o Dpkg::Options::="--force-confold" --no-install-recommends --no-install-suggests librdkafka-dev
RUN sudo apt-get install -y php-pear php-dev
RUN sudo pecl install rdkafka
RUN sudo echo -e "extension=rdkafka.so" > /etc/php/8.0/cli/php.ini
```
Now, i can use the rdkafka client in PHP.
Is it possible to set the PHP-Version in a dockerfile?

Beste regards
3xc3ption

**Comment by rfay** [comment]
You set the PHP version in ~/.ddev/config.yaml. `php_version: 8.0`

**Reply by 3xc3ption** [reply]
Yes, that is correct, but I was hoping that I can set a PHP version from the ~/.ddev/config.yaml via "ARG" or "ENV" in the .ddev/web-build/Dockerfile. 😃 

Beste regards
3xc3ption

**Comment by rfay** [comment]
So all supported versions of PHP are installed inside the container. What is it you want to accomplish?

**Reply by 3xc3ption** [reply]
Hi rfay,
i want the PHP_Version from the config.yaml into my dockerfile as an ARG, so that I avoid double settings.
Here is my finaly dockerfile, it works perfect with all PHP-Versions.
``` PHP
ARG BASE_IMAGE
FROM $BASE_IMAGE

ENV PHP_VERSION=7.3

RUN sudo apt-get update -y \
    && apt-get install -y php-pear php${PHP_VERSION}-dev \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y -o Dpkg::Options::="--force-confold" --no-install-recommends --no-install-suggests librdkafka-dev \
    && pear config-set php_suffix ${PHP_VERSION} \
    && pecl install rdkafka \
    && echo "extension=rdkafka.so" >> /etc/php/${PHP_VERSION}/cli/php.ini
```

**Comment by rfay** [comment]
So the image is built (build time) long before the PHP version is selected (runtime). So I understand why you want an arg to tell you this, but if you want that you can accomplish the same thing using a post-start hook, but it will slow down every single `ddev start`, whereas doing what's needed with a Dockerfile it's just once that it does this process.
